quinn-proto/src/lib.rs

```diff
@@ -68,6 +66,22 @@ pub use crate::streams::{FinishError, ReadError, WriteError};
 mod transport_error;
 pub use crate::transport_error::{Code as TransportErrorCode, Error as TransportError};
 
+/// Types that are generic over the crypto protocol implementation
+pub mod generic {
+    pub use crate::connection::Connection;
+    pub use crate::endpoint::{ClientConfig, Endpoint};
+    pub use crate::shared::ServerConfig;
+}
+
+/// A `Connection` using rustls for the cryptography protocol
+pub type Connection = generic::Connection<crypto::rustls::TlsSession>;
+/// A `ClientConfig` containing client-side rustls configuration
+pub type ClientConfig = generic::ClientConfig<crypto::rustls::ClientConfig>;
+/// An `Endpoint` using rustls for the cryptography protocol
+pub type Endpoint = generic::Endpoint<crypto::rustls::TlsSession>;
+/// A `ServerConfig` containing server-side rustls configuration
+pub type ServerConfig = generic::ServerConfig<crypto::rustls::ServerConfig>;
+
```

Comment:
Yeah, I was thinking we'd definitely want to allow compilation without rustls and/or ring.

Reply:
Maybe something like 

```rust
#[cfg(feature = "rustls")]
macro_rules! foo {
    {$($etc:tt)*} => { pub struct Foo<T = ()> { $($etc)* } };
}

#[cfg(not(feature = "rustls"))]
macro_rules! foo {
    {$($etc:tt)*} => { pub struct Foo<T> { $($etc)* } };
}

foo! {
    x: T
}
```

would work? It's not beautiful, but it's small.

Reply:
TBH I think that's making things much more complex... I'm also not wild aobut names having different meaning depending on the feature flags, I think error messages will likely make more sense if you keep them strictly separate.

Reply:
> I'm also not wild aobut names having different meaning depending on the feature flags

I think the example is consistent with the principle of feature flags, i.e. code written with the flag disabled should have identical semantics when the flag is enabled.

> I think error messages will likely make more sense if you keep them strictly separate.

Is a type alias strictly separate? I'd actually expect similar errors.for defaulted type parameters vs. type aliases with parameters supplied.

Reply:
So you're suggesting to have the root-exported names default to the rustls-based implementations if rustls is compiled in, or without default if not?

I'm not even sure if it's possible to set defaults in type aliases.

Reply:
Specifically I'm suggesting not using type aliases at all, but rather only exporting the types once, and defining them with defaults iff rustls is built in (if we actually make that dependency optional).

Reply:
So the reason I landed at this approach in the first place is that it prevents the abstract types from reaching the tests (and the other crate), which would make the commits substantially larger. As is, the change here proofs that the approach works, and from here we can have a bunch of options in how we expose the abstraction in our more outward facing APIs.

In my initial approach, I actually had a different approach in exposing the resulting types at the root level of the `-proto` crate, but this required a substantial number of changes in the test code. I find the current approach attractive as a way point since it proves that the abstraction works well in the type system without "infecting" a whole bunch of using code.

As such, I'd prefer not to change this as part of this PR, and then continue to iterate on the best way to expose this abstraction in future changes (which I'd be happy to work on).

Reply:
My expectation is that a defaulted type parameter should allow callers to be unmodified, but I'm content to leave this for future work.

Reply:
That was my expectation as well, but I have yet to get to work like that.